doc/dualtor/ICMP_Hardware_Offload_and_Protecion.md

```diff
@@ -0,0 +1,289 @@
+<div align="center"> <img src=image/sonic_component_interaction.png width=700 /> </div>
+
+### LinkMgrd
+With introduction of ICMP hardware offload a new config knob **[link_prober_type]** is added to differentiate between software and hardware based link probing. Hardware based Link Prober is added in LinkMgrd to support hardware based probing that produces the ICMP echo sessions entries in App DB ICMP_ECHO_SESSION_TABLE that gets consumed by orchagent. LinkMgrd will consume icmp echo session entries from State DB ICMP_ECHO_SESSION_TABLE to determine the hardware prober state and trigger events for Link state machines.
+
+#### Session cookie
+Session cookie is used by NPU to distinguish between software ICMP packets and ICMP packets that needs to be handled by the offload engine in NPU. LinkMgrd will use separate cookies for software and hardware based sessions.
+
+#### Session GUID handling
+Session GUID is used to distinguish between the ICMP sessions. However session GUID is not generated per session in the current implementation, yet it works as the software based probing uses raw Linux packet socket on the interface. This way software does not need to demultiplex icmp echo reply packets for each link's session based on session GUID. However in hardware based approach NPU uses the session GUID to map the icmp packets to a session. So with introduction of hardware probing session GUID will always be a randomly generated unique value for both software and hardware sessions.
+
+To avoid collision of GUIDs between self and peer sessions, last byte of session GUID will be set to a fixed value unique to a ToR. ToR roles like Upper ToR and Lower ToR can be determined based on multiple factors and Linkmgrd can as well use some preset values based on these roles as the last byte of session GUID.
+
+#### Peer session handling
+Hardware Link prober needs to know the peer session GUID to set the peer session entry in App DB. For this the hardware link prober in LinkMgrd will listen / receive initially icmp echo reply packets from peer over the Linux packet socket on the interface, reusing the mechanism used in software link prober. It will learn the session GUID from the payload of these icmp echo reply packets from peer and use this learnt session GUID to create the hardware offload session entry for peer in App DB.
+
+#### Timer considerations
+* **Minimum timer value:** Minimum allowed timer value for software session will be set to 100 msec and for hardware session will be 3 msec.
+* **ICMP echo session tx_interval setting:**
+    * tx_interval = probing interval
+    * Positive probing timer value = probing interval x positive signal count.
+    * LinkMgrd will start positive probing timer after receiving the first UP state notification for a session from State DB and link prober will conitnue to remain in unknown state until the expiry of this positive timer at which it will transition the Link prober to Active state.

```

Comment:
s/conitnue/continue

Reply:
So for active-active dualtor, after the state machine init and ICMP echo session creation, the link prober remains in `unknown`, it will try to toggle to standby in this case; and after the positive timer expires, link prober recovers back to active -> toggle to active; It appears that there will be an extra toggle in the mux port init stage (this is not expected).